src/ppopt/utils/region_overlap_utils.py

```diff
@@ -0,0 +1,188 @@
+    regions = [*regions, *new_regions]
+    # regions[0] == to_remove[0]
+    # purge
+    regions = [cr for cr in regions if cr not in to_remove]
+
+    return possible_dual_degeneracy, regions
+
+
+def adjust_regions_at_intersection(new_regions: List[CriticalRegion], inner_region: CriticalRegion,
+                                   outer_region: CriticalRegion,
+                                   intersection: float, expand_outer_on_left: bool) -> Tuple[
+    List[CriticalRegion], CriticalRegion, CriticalRegion]:
+    new_regions.append(copy.deepcopy(outer_region))
+    inner_lb, inner_ub = get_bounds_1d(inner_region.E, inner_region.f)
+    if expand_outer_on_left:
+        outer_region = cr_new_bounds(outer_region, lb_new=None, ub_new=intersection)
+        inner_region = cr_new_bounds(inner_region, lb_new=intersection, ub_new=None)
+        new_regions[-1] = cr_new_bounds(new_regions[-1], lb_new=inner_ub, ub_new=None)
+    else:
+        outer_region = cr_new_bounds(outer_region, lb_new=None, ub_new=inner_lb)
+        inner_region = cr_new_bounds(inner_region, lb_new=None, ub_new=intersection)
+        new_regions[-1] = cr_new_bounds(new_regions[-1], lb_new=intersection, ub_new=None)
+    return new_regions, inner_region, outer_region
+
+
+def split_outer_region(new_regions: List[CriticalRegion], cr: CriticalRegion, inner_lb: float, inner_ub: float) -> \
+        Tuple[List[CriticalRegion], CriticalRegion]:
+    new_regions.append(copy.deepcopy(cr))
+    cr = cr_new_bounds(cr, lb_new=None, ub_new=inner_lb)
+    new_regions[-1] = cr_new_bounds(new_regions[-1], lb_new=inner_ub, ub_new=None)
+    return new_regions, cr
+
+
+def cr_new_bounds(cr: CriticalRegion, lb_new: Optional[float], ub_new: Optional[float]) -> CriticalRegion:
+    lb, ub = get_bounds_1d(cr.E, cr.f)
+
+    lb = lb if lb_new is None else lb_new
+    ub = ub if ub_new is None else ub_new
+
+    cr.E = numpy.array([[1], [-1]])
+    cr.f = numpy.array([[ub], [-lb]])
+
+    return cr
+
+
+def full_overlap(cr1: CriticalRegion, cr2: CriticalRegion) -> bool:
+    # region 2 fully inside region 1
+    lb1, ub1 = get_bounds_1d(cr1.E, cr1.f)
+    lb2, ub2 = get_bounds_1d(cr2.E, cr2.f)
+    return lb1 <= lb2 and ub1 >= ub2
+
+
+def partial_overlap(cr1: CriticalRegion, cr2: CriticalRegion) -> bool:
+    # region 1 to the left of region 2
+    lb1, ub1 = get_bounds_1d(cr1.E, cr1.f)
+    lb2, ub2 = get_bounds_1d(cr2.E, cr2.f)
+    return lb1 < lb2 and ub1 > lb2 and ub2 > ub1
+
+
+def equal_linear_objective(program: MPMILP_Program, inner_region: CriticalRegion, outer_region: CriticalRegion) -> bool:
+    f_inner_lb, f_inner_ub, f_outer_lb, f_outer_ub = evaluate_objective_at_middle_bounds(program, inner_region,
+                                                                                         outer_region)
+    return f_inner_lb == f_outer_lb and f_inner_ub == f_outer_ub
+
+
+def region_2_dominates(program: MPMILP_Program, cr_1: CriticalRegion, cr_2: CriticalRegion) -> bool:
```

Comment:
This is duplicate logic with the next function, e.g. we can collapse both of them into a single function e.g. ``region_dominates(program, cr1, cr2) & region_dominates(program, cr2, cr1) ``